alicat/util.py

```diff
@@ -120,16 +117,17 @@ async def _handle_communication(self, command: str) -> str | None:
             return None
 
     async def _handle_connection(self) -> None:
-        """Automatically maintain TCP connection."""
-        if self.open:
-            return
-        try:
-            await asyncio.wait_for(self._connect(), timeout=0.75)
-            self.reconnecting = False
-        except (asyncio.TimeoutError, OSError):
-            if not self.reconnecting:
-                logger.error(f'Connecting to {self.address} timed out.')
-            self.reconnecting = True
+        """Automatically maintain connection."""
+        async with self.lock:
```

Comment:
did you find adding the lock improved anything?

Similar code used to have a lock but there was no apparent downside to removing it: https://github.com/numat/productivity/pull/80/files.

Reply:
I found this lock was critical for my use case..  I've got two coroutines that are attempting to connect to two different FlowControllers on the same serial device, around about the same time.  Attempting to connect to both without this lock leads to corruption of the serial stream.

Unfortunately, introducing this lock causes some issues, and is actually why I removed all of the `_handle_connection` calls in the `_read` and `_write` methods..  With the lock acquired for connection, `_handle_communication` is called which calls `_write` which calls `_handle_connection` which deadlocks because asyncio Locks aren't re-entrant.